core/corehttp/gateway_handler_unixfs__redirects.go

```diff
@@ -0,0 +1,289 @@
+// doesn't exist.  However, a `_redirects` file may exist and contain a redirect rule that redirects that path to a different path.
+// We need to evaluate the rule and perform the redirect if present.
+//
+// Example 3:
+// Another possibility is that the path corresponds to a rewrite rule (i.e. a rule with a status of 200).
+// In this case, we don't perform a redirect, but do need to return a `path.Resolved` and `path.Path` corresponding to
+// the rewrite destination path.
+//
+// Note that for security reasons, redirect rules are only processed when the request has origin isolation.
+func (i *gatewayHandler) handleUnixfsPathResolution(w http.ResponseWriter, r *http.Request, contentPath ipath.Path, logger *zap.SugaredLogger) (ipath.Resolved, ipath.Path, bool) {
+	// Attempt to resolve the provided path.
+	resolvedPath, err := i.api.ResolvePath(r.Context(), contentPath)
+
+	switch err {
+	case nil:
+		return resolvedPath, contentPath, true
+	case coreiface.ErrOffline:
+		webError(w, "ipfs resolve -r "+debugStr(contentPath.String()), err, http.StatusServiceUnavailable)
+		return nil, nil, false
+	default:
+		// The path can't be resolved.
+		// If we have origin isolation, attempt to handle any redirect rules.
+		if hasOriginIsolation(r) {
+			redirectsFile := i.getRedirectsFile(r, logger)
+			if redirectsFile != nil {
+				redirectRules, err := i.getRedirectRules(r, redirectsFile)
+				if err != nil {
+					internalWebError(w, err)
+					return nil, nil, false
+				}
+
+				redirected, newPath, err := i.handleRedirect(w, r, redirectRules)
+				if err != nil {
+					err = fmt.Errorf("trouble processing _redirects file at %q: %w", redirectsFile.String(), err)
+					internalWebError(w, err)
+					return nil, nil, false
+				}
+
+				if redirected {
+					return nil, nil, false
+				}
+
+				// 200 is treated as a rewrite, so update the path and continue
+				if newPath != "" {
+					// Reassign contentPath and resolvedPath since the URL was rewritten
+					contentPath = ipath.New(newPath)
+					resolvedPath, err = i.api.ResolvePath(r.Context(), contentPath)
+					if err != nil {
+						internalWebError(w, err)
+						return nil, nil, false
+					}
+
+					return resolvedPath, contentPath, true
+				}
+			}
+		}
+
+		// if Accept is text/html, see if ipfs-404.html is present
+		// This logic isn't documented and will likely be removed at some point.
+		// Any 404 logic in _redirects above will have already run by this time, so it's really an extra fall back
+		if i.servePretty404IfPresent(w, r, contentPath) {
+			logger.Debugw("serve pretty 404 if present")
+			return nil, nil, false
+		}
+
+		// Fallback
+		webError(w, "ipfs resolve -r "+debugStr(contentPath.String()), err, http.StatusNotFound)
+		return nil, nil, false
+	}
+}
+
+func (i *gatewayHandler) handleRedirect(w http.ResponseWriter, r *http.Request, redirectRules []redirects.Rule) (bool, string, error) {
```

Comment:
We already do a lot of redirects unrelated to `_redirects`, let's rename to make it clear what happens here:
```suggestion
func (i *gatewayHandler) handleRedirectsFileRules(w http.ResponseWriter, r *http.Request, redirectRules []redirects.Rule) (bool, string, error) {
```